### PR TITLE
Extracted Loader Classes into separate files

### PR DESCRIPTION
**Pull request by dyt811**
#### GitHub

- [X] I've given this PR a concise, self-descriptive, and meaningful title
- [X] I've linked relevant issues in the PR body
- [X] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [X] I've assigned a reviewer

#### PR contents

- [X] I've consulted [ivadomed's internal developer documentation](https://github.com/ivadomed/ivadomed/wiki) to ensure my contribution is in line with any relevant design decisions
- [X] I've added [relevant tests](https://github.com/ivadomed/ivadomed/wiki/tests) for my contribution
- [X] I've updated the [relevant documentation](https://github.com/ivadomed/ivadomed/wiki/documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description

Refactoring proceeding to the multi-session/contrasts revamp
- For every new file created, **no file content change**. Utilized PyCharm's refactoring function to move the files into respective modules, that is all. 
- Added inline import where needed to bypass circular import

## Linked issues

- Necessary file reorg/refactor for multi-session/contrasts revamp to help focus problems/changes make to upcoming BIDS related sections.

### CODE REVIEW

**Review by mariehbourget — COMMENTED** [review]
Refactoring is very clean, I left a few comments for minor corrections.
As discussed in meeting, I don't foresee any compatibility issues between this refactoring and future loader's endeavor.

I have more of a general question about the separation of `loader.py` into several files. Is this standard practice to have only 1 function/class per file? I don't have an issue or strong opinion either way but I'm curious about best practice and design choice. Could there be a reason to group some classes together or is it better to have everything separate?

I also don't know much about circular/inline imports and cannot give an informed feedback on this.

I think the PR should also be reviewed by someone more on the soft dev side.

**Comment by dyt811** [comment]
> I have more of a general question about the separation of `loader.py` into several files. Is this standard practice to have only 1 function/class per file? I don't have an issue or strong opinion either way but I'm curious about best practice and design choice. Could there be a reason to group some classes together or is it better to have everything separate?

So usually, we want to keep each python module files as unitarily small as possible. Typically class are suppose to stand alone in class specific modules or files etc. The reason this is done to mainly reduce the cognitive load on programmer when working with files/modules/class. Typically, with modern IDE like PyCharm, you can easily auto link to functions/classes across files (which already has hopefully informative names) but scrolling to a few hundred or more often thousand line multi class/multimodule is very challenging to consistent navigate successfully. As far as I know, this is the biggest reason typically people try to keep module/class/files small as possible and separate out specific responsibilities out. 

Also, file/module/class names typically give people a limited scope/hint about what the changes are. Loader.py is a VERY big scope comparing to making changes to BIDSDataset.py for instance.

The only potentially possible benefit I have seen with keeping multiple classes/functions together is to not have to worry about potential import issues by allowing all classes within the same file to reference each other with almost impunity. This is actually a con in a way, by making this easy, there could be introduce implicit dependencies across classes that are not obvious. Forcing separate classes out, make this VERY clear and explicit. 

> I also don't know much about circular/inline imports and cannot give an informed feedback on this.

So this is the reason I believe whoever wrote the original class was potentially trying to avoid (by putting them into a single file). If we simply breaking the class out without testing while import loader.py function, it could depend on A which depend on loader.py introduces circular dependency. 

Circular dependency typically happens when there is some sort of architecture issue. 

> I think the PR should also be reviewed by someone more on the soft dev side.

Yup. Good point. I will try to see if any of the new folks coming on boarding can share any good insight as well.

File: testing/unit_tests/test_loader.py [hunk 1]
```diff
@@ -3,6 +3,8 @@
 import csv_diff
 import logging
 import torch
+
+from ivadomed.loader.utils import dropout_input
```
[the remark this answers is not in the document]
Comment (dyt811) [reply testing/unit_tests/test_loader.py:7]: Very valid point. I have adjusted accordingly. A future PR will adjust the loader until functions as well so there will be more change/grouping potentially.

File: ivadomed/loader/loader.py [hunk 2]
```diff
@@ -53,6 +41,8 @@ def load_dataset(bids_df, data_list, transforms_params, model_params, target_suf
     Note: For more details on the parameters transform_params, target_suffix, roi_params, contrast_params,
     slice_filter_params and object_detection_params see :doc:`configuration_file`.
     """
+    from ivadomed.loader.bids3d_dataset import Bids3DDataset
+    from ivadomed.loader.bids_dataset import BidsDataset
```
[the remark this answers is not in the document]
Comment (dyt811) [reply ivadomed/loader/loader.py:45]: Hmm... Great spot! I thought this was causing circular import but it apparently doesn't because base on the latest check. Moving it to the top then in this case.

Typically, import is done at module level such that it is only importing it once and won't have to do it at functional level. 

Sometimes, structural/architecture reason arises, that if module functionalities are not properly segregated and have dependencies on each other, it will literally throw a partial module initialized error. I was mistaken in this case so we should move it to the top. Corrected.

File: ivadomed/object_detection/utils.py [hunk 3]
```diff
@@ -108,6 +106,7 @@ def generate_bounding_box_file(subject_path_list, model_path, path_output, gpu_i
         dict: Dictionary containing bounding boxes related to their image.
 
     """
+    from ivadomed import inference as imed_inference
```
[the remark this answers is not in the document]
Comment (dyt811) [reply ivadomed/object_detection/utils.py:109]: Here is a an example of what happens if this gets moved up top, causing circular import. 
image

**Comment by mariehbourget** [comment]
Thank you for the detailed explanation @dyt811, it is much appreciated, I'll keep that in mind for future implementation :)

File: ivadomed/object_detection/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (copperwiring) [reply ivadomed/object_detection/utils.py:109]: @dyt811 I see that the circular import may be breaking the test. However, instead of  `from ivadomed import inference as imed_inference` why can we write `import ivadomed.inference as imed_inference` at the top.

From my understanding, what do you is not wrong but then is there a set way it has been done is the past. We can kep absolute import bound to a name at the top or move it later. Unless we don't have too many later imports, I don't see why what you have done can be a problem. If we do come across too many such (later) imports, it may affect code clarity.

**Review by copperwiring — CHANGES_REQUESTED** [review]

**Comment by copperwiring** [comment]
It appears that there may be new changes in the master branch which can prevent this merge (`This branch is out-of-date with the base branch`). Can you update this branch with any current changes too?

**Comment by dyt811** [comment]
> However, instead of `from ivadomed import inference as imed_inference` can we write `import ivadomed.inference as imed_inference` at the top?

So this apparently did not change anything per highlighted changed line below.  😢 

image

BTW, you can test these as well when using `pytest` against the particular unit test file, in this case, the `/testing/unit_tests/test_orientation.py` file

**Comment by copperwiring** [comment]
This is to confirm that all changes are good to go and unit tests are working correctly.

**Review by copperwiring — APPROVED** [review]
